File: apps/api/app/schemas/lesson.py
from typing import Literal

from app.schemas.common import ApiModel
from app.schemas.equation import MathValue, QuadraticCoefficients

SolutionMethod = Literal["factoring", "square_root", "completing_the_square", "quadratic_formula"]
LessonStatus = Literal["completed", "unsupported_instructional_method"]


class MathLine(ApiModel):
    id: str
    expression: str
    latex: str


class TeachingStep(ApiModel):
    id: str
    title: str
    step_type: str
    math_lines: list[MathLine]


class LessonResponse(ApiModel):
    status: LessonStatus
    original_equation: str
    normalized_equation: str
    method: SolutionMethod | None
    coefficients: QuadraticCoefficients
    solutions: list[MathValue]
    steps: list[TeachingStep]
    unsupported_reason: str | None = None

File: packages/types/src/lesson.ts
export type SolutionMethod =
  | "factoring"
  | "square_root"
  | "completing_the_square"
  | "quadratic_formula";

export type LessonStatus = "completed" | "unsupported_instructional_method";

export interface MathValue {
  expression: string;
  latex: string;
}

export interface QuadraticCoefficients {
  a: MathValue;
  b: MathValue;
  c: MathValue;
}

export interface MathLine {
  id: string;
  expression: string;
  latex: string;
}

export interface TeachingStep {
  id: string;
  title: string;
  stepType: string;
  mathLines: MathLine[];
}

export interface Lesson {
  status: LessonStatus;
  originalEquation: string;
  normalizedEquation: string;
  method: SolutionMethod | null;
  coefficients: QuadraticCoefficients;
  solutions: MathValue[];
  steps: TeachingStep[];
  unsupportedReason?: string;
}
